dockstore-webservice/src/main/resources/migrations.1.7.0.xml

```diff
@@ -159,4 +159,43 @@
         <setColumnRemarks columnName="lastupdated" tableName="tool" remarks="For automated builds: last time tool/namespace was refreshed Dockstore, tool info updated, default version selected. For hosted tools: when you created the tool"/>
         <setColumnRemarks columnName="lastupdated" tableName="workflow" remarks="For remote: When refresh all is hit for first time. Hosted: Time created."/>
     </changeSet>
+    <changeSet id="dyuen" author="frozen">
+        <addColumn tableName="tag">
+            <column defaultValueBoolean="false" name="frozen" type="bool"/>
+        </addColumn>
+        <addColumn tableName="workflowversion">
+            <column defaultValueBoolean="false" name="frozen" type="bool"/>
+        </addColumn>
+        <addColumn tableName="sourcefile">
+            <column defaultValueBoolean="false" name="frozen" type="bool"/>
+        </addColumn>
+        <sql dbms="postgresql">
+            /* lock down workflowversion table, even for dockstore user as table owner  */
+            alter table workflowversion enable row level security;
+            alter table workflowversion force row level security;
+            /* all users can read the table, but cannot update or delete frozen versions */
+            create policy select_frozenversions on workflowversion for select using (true);
+            create policy insert_frozenversions on workflowversion for insert with check (true);
+            create policy update_frozenversions on workflowversion for update using (not frozen) with check (true);
+            create policy delete_frozenversions on workflowversion for delete using (not frozen);
+
+            /* lock down workflowversion table, even for dockstore user as table owner  */
+            alter table tag enable row level security;
+            alter table tag force row level security;
+            /* all users can read the table, but cannot update or delete frozen versions */
+            create policy select_frozenversions on tag for select using (true);
+            create policy insert_frozenversions on tag for insert with check (true);
+            create policy update_frozenversions on tag for update using (not frozen) with check (true);
+            create policy delete_frozenversions on tag for delete using (not frozen);
```

Comment:
What if we do need to delete a version, e.g., somebody accidentally put credentials in a frozen version or something like that?

Can the sysadmin still delete it without changing the policies, or will the sysadmin have to disable this policy?

Reply:
db superusers can delete, i think other users including owner would need to disable the policy unless we changed it